src/knitlib/communication/communication.py

```diff
@@ -0,0 +1,59 @@
+# This file is based on Ayab
+
+"""Handles the serial communication protocol.
+"""
+
+import time
+import serial
+
+import logging
+
+
+class Communication(object):
+  """Class Handling the serial communication protocol."""
+
+  def __init__(self, serial=None):
+    """Creates an Communication object, with an optional serial-like object."""
+    logging.basicConfig(level=logging.DEBUG)
+    self.__logger = logging.getLogger(__name__)
+    self.__ser = serial
+
+  def __del__(self):
+    """Handles on delete behavior closing serial port object."""
+    self.close_serial()
+
+  def open_serial(self, pPortname=None):
+    """Opens serial port communication with a portName."""
+    if not self.__ser:
```

Comment:
This should have an else. Missed that when writing:  https://github.com/tian2992/AYAB/blob/master/software/python/ayab/plugins/ayab_plugin/ayab_communication.py#L48 :disappointed: